Add mock Segment server

Gudahtt opened the pull request:
This mock Segment server can be used to test our extension metrics. It will respond to all request with HTTP 200, and will print the requests to the console. It also has parsing built-in for Segment request payloads.

Right now only the event name is printed, but we can enhance this in the future to print more event information. We can also enhance the mock to be a more realistic representation of the API.

The extension has been modified to allow the Segment host to be overwritten with the `SEGMENT_HOST` environment variable. This will ensure that all Segment events are redirected to that host.

So for example, to create a dev build that uses this server, you could set the `SEGMENT_WRITE_KEY` and `SEGMENT_LEGACY_WRITE_KEY` values to any non-empty string, and set `SEGMENT_HOST` to `http://localhost:9090`.

This was created originally to test PR #9768

Gudahtt committed:
Add mock Segment server

This mock Segment server can be used to test our extension metrics. It
will respond to all request with HTTP 200, and will print the requests
to the console. It also has parsing built-in for Segment request
payloads.

Right now only the event name is printed, but we can enhance this in
the future to print more event information. We can also enhance the
mock to be a more realistic representation of the API.

The extension has been modified to allow the Segment host to be
overwritten with the `SEGMENT_HOST` environment variable. This will
ensure that all Segment events are redirected to that host.

So for example, to create a dev build that uses this server, you could
set the `SEGMENT_WRITE_KEY` and `SEGMENT_LEGACY_WRITE_KEY` values to
any non-empty string, and set `SEGMENT_HOST` to
`http://localhost:9090`.

This was created originally to test PR #9768

darkwing commented on development/build/scripts.js:
```diff
@@ -372,6 +373,7 @@ function createScriptTasks({ browserPlatforms, livereload }) {
         INFURA_PROJECT_ID: opts.testing
           ? '00000000000000000000000000000000'
           : conf.INFURA_PROJECT_ID,
+        SEGMENT_HOST: conf.SEGMENT_HOST ? conf.SEGMENT_HOST : undefined,
```
Can we do `conf.SEGMENT_HOST ?? undefined`, or just `conf.SEGMENT_HOST`?

Gudahtt replied:
I had tried to use `conf.SEGMENT_HOST ?? undefined` initially, but this script isn't run through `babel` so it failed (`??` isn't supported by Node v10).

I suppose just `conf.SEGMENT_HOST` would work as well :thinking: I wanted to ensure this was left `undefined` rather than as an empty string if it wasn't set. I'll check to see what the default is - I never tested it.

Gudahtt replied:
It was unnecessary after all! It defaults to `undefined` if it's not set. I have removed this ternary in https://github.com/MetaMask/metamask-extension/pull/9783/commits/9d5a838abfdea196c8f73f1288031cf30518ce63

Gudahtt committed:
Remove redundant ternary

darkwing commented:
When attempting to use this utility, whenever a segment event is mean to be sent, I get the following error:

```
(node:8227) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at Server.http.createServer (/Users/davidwalsh/Projects/metamask-extension/development/lib/create-segment-server.js:48:33)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:8227) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 26)
```

One way to trigger this issue is clicking this button:

<img width="739" alt="IAgree" src="https://user-images.githubusercontent.com/46655/98389827-34159400-201a-11eb-84d7-acc033bc0280.png">

There appears to be no request body:

<img width="1177" alt="Nada" src="https://user-images.githubusercontent.com/46655/98389858-3c6dcf00-201a-11eb-9fb3-c66d10c70ac9.png">

Tested in Firefox.

Gudahtt committed:
Handle preflight requests and empty body

Preflight requests are now handled by replying that all request
methods, headers, and origins are allowed. We could be more specific,
but it didn't seem necessary.

Handling preflight requests was necessary because Firefox considers
`localhost` to be a cross-origin resource.

The mock Segment server will now also handle requests with an empty
body. It will return an empty array for the `metricEvents` parameter
instead of blowing up.

Gudahtt commented:
Thanks @darkwing! That was my bad - I made the mistake of only testing in Chrome, not in Firefox.

It turns out that Firefox sends a preflight request before initially sending anything to `localhost`, which this mock server wasn't handling. I also wasn't handling empty requests in general. Both issues have been addressed in https://github.com/MetaMask/metamask-extension/pull/9783/commits/91fa3c226a8afe8a7bd6c220ef78dd456763d064

darkwing reviewed (approved):
Excellent, works for me!  Would we also want to spit out data, or just the event type?

<img width="518" alt="Screenshot at Nov 09 10-06-14" src="https://user-images.githubusercontent.com/46655/98565417-36bef600-2273-11eb-99e3-676a72a2e227.png">

Gudahtt commented:
>Would we also want to spit out data, or just the event type?

Yes, I think this would be very useful as well, and fairly easy to do. I'll add this in a separate PR
